Add support for policy objects

qha opened the pull request:
Fixes: #316

TjeuKayim commented on lib/puppet/type/firewalld_policy.rb:
```diff
@@ -0,0 +1,308 @@
+        raise Puppet::Error, 'egress egress_zones must contain at least one zone'
+      end
+
+      return if value.length == 1
```
`value` can be a string here if passed `['ANY']`, and the length will be the string length instead of the array length. This check does not work as expected.

qha replied:
Fixed in recently pushed commit but not tested beyond the spec tests and not rebased.

TjeuKayim reviewed (commented):
I tested this pull-request, and it is functional. Only minor issues that I could work around. Good how you made it fit in nicely with the rest of the module.

qha commented:
Thank you for your review! I will work on the issues you spotted but it might take a few weeks as i've dug down in some other projects...

qha commented on lib/puppet/type/firewalld_policy.rb:
```diff
@@ -0,0 +1,308 @@
+  newproperty(:priority) do
+    desc 'The priority of the policy as an integer (default -1)'
+
+    defaultto('-1')
```
Fixed in recently pushed commit but not tested beyond the spec tests and not rebased.

qha commented:
As you might have seen i've addressed the issues you spotted but i won't have time to test them properly until some time in august at best.

qha commented:
I just did some real life testing and things seem to work as expected.

I've squished my fixes and rebased on master to remove the merge conflict.

After rebasing the spec tests report an error on my workstation but i don't think it's related as the same error is present in the master branch.

qha commented:
Sorry, i need to fix another thing. The firewalld_policy type insists that ingress and egress zones are arrays which is normally resonable but i don't think it is when ensure is absent.

qha commented:
There, ingress_zone and egress_zone may now be emtpy if ensure is absent.

jcpunk commented:
Can I help get this ready for merge in?

jcpunk commented:
@qha can you squash this down again?

qha commented:
Sorry for being so slow to respond, i was busy with going to and recovering from a conference.
If i'm supposed to squash my changes into some other combination of commits i would need more detailed specifications as i feel the current partitioning is the most appropriate and the remote second should be available to whoever merges this with two clicks (squash and merge).
@jcpunk

jcpunk commented:
The main squashing I had in mind was regarding the readme and reference updates.

qha commented:
I see. If i recall correctly i did it that way because the reference had not been updated the last few commits so updating it introduced unrelated changes that i did not want to put in the same commit.
Would it help if i updated the reference before the main commit in this pr in one commit and then did it again as part of the main commit?

qha commented on lib/puppet/type/firewalld_policy.rb:
```diff
@@ -0,0 +1,325 @@
+    end
+  end
+
+  def purge_rich_rules
```
Perhaps i would have known while this was fresh in my mind but i don't now. I hope someone else can answer this.

jcpunk commented:
> I see. If i recall correctly i did it that way because the reference had not been updated the last few commits so updating it introduced unrelated changes that i did not want to put in the same commit. Would it help if i updated the reference before the main commit in this pr in one commit and then did so again as part of the main commit?

That is fine with me

jameslikeslinux commented on lib/puppet/provider/firewalld_port/firewall_cmd.rb:
```diff
@@ -11,7 +11,16 @@
+                                           true, false)
+             else
+               execute_firewall_cmd(['--query-port', @rule_args],
+                                    @resource[:policy],
```
This should be `@resource[:zone]`

qha replied:
Good catch, thank you!

qha commented:
> > I see. If i recall correctly i did it that way because the reference had not been updated the last few commits so updating it introduced unrelated changes that i did not want to put in the same commit. Would it help if i updated the reference before the main commit in this pr in one commit and then did so again as part of the main commit?
> 
> That is fine with me

I can't find the other changes i thought i remembered so i just squashed the last commit into the main one.

jcpunk commented:
Unless there are any objections I'd like to merge this in next week.
